### ISSUE TITLE
v7.7 build failures on Windows using VS2017 and VS2019

### ISSUE THREAD

**Issue by oleurodecision** [google/or-tools#2083]
**What version of OR-tools and what language are you using?**
Version: v7.7
Language: C++

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
Default ones for now.

**What operating system (Linux, Windows, ...) and version?**
Windows Server 2016 / Windows 10 Professional, both 64 bits

**What did you do?**
see joined batch script for details, main steps :
- clone the repository
- set up MSVC environment
  - VS 2017 v15.9.24 on the WS2016 system
  - VS 2019 Community on the W10Pro system
- try to build using both
  - the given commands (https://developers.google.com/optimization/install/cpp/source_windows) 
  - the cmake scripts (https://github.com/google/or-tools#experimental-build-with-cmake)

**What did you expect to see**
successful build :)

**What did you see instead?**
- with WS2016/VS2017/cmake and W10Pro/VS2019/cmake some static assertion failure (same as #2077, which was closed for - I think - wrong reason)
- with WS2016/VS2017/msvc build failure due to internal compiler error
- with W10Pro/VS2019/msvc segfault with running min_cost_flow unit test

[build.vs2017.cmake.log](https://github.com/google/or-tools/files/4835787/build.vs2017.cmake.log)
[build.vs2017.msvc.log](https://github.com/google/or-tools/files/4835788/build.vs2017.msvc.log)
[build.vs2019.cmake.log](https://github.com/google/or-tools/files/4835789/build.vs2019.cmake.log)
[build.vs2019.msvc.log](https://github.com/google/or-tools/files/4835790/build.vs2019.msvc.log)

As a side note, the VS2017 build is my "only" targeted build type, the VS2019 was tested due to the VS2017 failures. 
Thanks for any hint on my mistakes or how to help fix the issue

**Comment by lperron** [comment]
Vs2017 is no longer supported.

We will check the vs 2019 failure.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le ven. 26 juin 2020 à 10:15, oleurodecision <notifications@github.com> a
écrit :

**Comment by gregy4** [comment]
VS2017 is compatible with ortools-7.7 but you should use clang compiler. I successfully build windows version with bazel as described in issue #2072.

**Comment by philsuess** [comment]
In hopes that it might help, I have the same problem when I try to build a debug version using VS2019 (and conan on top). Here is what conan uses:
```
[settings]
    arch=x86_64
    build_type=Debug
    compiler=Visual Studio
    compiler.runtime=MDd
    compiler.version=16
    os=Windows
```

and here is the relevant cmake output (cmake command is shown in last line of this excerpt)
```
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.17134.
-- ortools version: 7.8.7959
-- Build C++ library: ON
-- Build Python: OFF
-- Build Java: OFF
-- Build .Net: OFF
-- Build samples: ON
-- Build examples: OFF
-- Build all dependencies: ON
-- Build ZLIB: ON
-- Build abseil-cpp: ON
-- Build gflags: ON
-- Build glog: ON
-- Build protobuf: ON
-- SCIP support: ON
-- Build SCIP: ON
-- COIN-OR support: ON
-- Build CoinUtils: ON
-- Build Osi: ON
-- Build Clp: ON
-- Build Cgl: ON
-- Build Cbc: ON
-- CPLEX support: OFF
-- XPRESS support: OFF
-- Fetching ZLIB
-- Fetching ZLIB - fetched
-- Fetching Abseil-cpp
--   No CMAKE_CXX_STANDARD set, assuming 11
-- Fetching Abseil-cpp - fetched
-- Fetching gflags
-- Fetching gflags - fetched
-- Fetching glog
-- Fetching glog - fetched
-- Fetching SCIP
--   Build shared libraries: OFF
--   Build type: Release
--   Finding ZLIB
--   Finding ZLIB - found
--   Support Readline: OFF
--   Support GMP: OFF
--   Support PAPILO: OFF
--   Finding Solver "none"
--   Support SOPLEX: OFF
--   Support CLP: OFF
--   Support CPLEX: OFF
--   Support GLOP: OFF
--   Support GUROBI: OFF
--   Support XPRESS: OFF
--   Support MOSEK: OFF
--   Support QSO: OFF
--   Finding symmetry computation program "none"
--   Support SYM: OFF
--   CodeCoverage: OFF
--   Support ZIMPL: OFF
--   Support IPOPT: OFF
--   Support WORHP: OFF
--   Finding CRITERION
--   Could NOT find CRITERION (missing: CRITERION_LIBRARY CRITERION_INCLUDE_DIR) 
--   Finding CRITERION - not found
--   Support AMPL: OFF
-- Fetching SCIP - fetched
-- Fetching Protobuf
--   Building Protobuf: ...
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.17134.
-- Configuring done
-- Generating done
-- Build files have been written to: R:/Radiotherapy/reps/google-or-tools/or-tools/_deps/protobuf-subbuild/project_build
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Framework

(...)

D:\temp\google-or-tools\or-tools\ortools/sat/encoding.h(131,36): error C2338: ERROR_EncodingNode_is_not_well_compacted (compiling source file D:\temp\google-or-tools\or-tools\ortools\bop\complete_optimizer.cc) [D:\temp\google-or-tools\or-tools\ortools\bop\ortools_bop.vcxproj]

D:\temp\google-or-tools\or-tools\ortools/sat/encoding.h(131,36): error C2338: ERROR_EncodingNode_is_not_well_compacted (compiling source file D:\temp\google-or-tools\or-tools\ortools\sat\encoding.cc) [D:\temp\google-or-tools\or-tools\ortools\sat\ortools_sat.vcxproj]

D:\temp\google-or-tools\or-tools\ortools/sat/encoding.h(131,36): error C2338: ERROR_EncodingNode_is_not_well_compacted (compiling source file D:\temp\google-or-tools\or-tools\ortools\bop\bop_portfolio.cc) [D:\temp\google-or-tools\or-tools\ortools\bop\ortools_bop.vcxproj]

D:\temp\google-or-tools\or-tools\ortools/sat/encoding.h(131,36): error C2338: ERROR_EncodingNode_is_not_well_compacted (compiling source file D:\temp\google-or-tools\or-tools\ortools\bop\bop_solver.cc) [D:\temp\google-or-tools\or-tools\ortools\bop\ortools_bop.vcxproj]

D:\temp\google-or-tools\or-tools\ortools/sat/encoding.h(131,36): error C2338: ERROR_EncodingNode_is_not_well_compacted (compiling source file D:\temp\google-or-tools\or-tools\ortools\sat\optimization.cc) [D:\temp\google-or-tools\or-tools\ortools\sat\ortools_sat.vcxproj]

conan : ERROR: conanfile.py (ortools/7.8): Error in build() method, line 42
At line:1 char:1
+ conan build . 2>&1 > conan_build_log
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (ERROR: conanfil...method, line 42:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
	cmake.build(target='install')
	ConanException: Error 1 while executing cmake --build D:\temp\google-or-tools/or-tools --config Debug --target install
```

**Comment by Mizux** [comment]
@philsuess your issue is an other issue when using Debug in a CMake based build on windows...
Also spotted here: https://ci.appveyor.com/project/Mizux/cmake-ortools/branch/master

edit: Just create the issue #2147

**Comment by philsuess** [comment]
@Mizux: thanks. I subbed your new issues and will be watching closely and eagerly. 

I thought the issue with the static assert is valid here, as OP pointed out has the same issue with a vs2019/CMake Build (albeit not a debug one).

**Comment by Mizux** [comment]
Should be fixed on master (see #2147) -> I'll close it.
